panel/io/application.py

```diff
@@ -17,19 +21,63 @@
 from bokeh.application.handlers.document_lifecycle import (
     DocumentLifecycleHandler,
 )
+from bokeh.application.handlers.function import FunctionHandler
+from bokeh.models import CustomJS
 
 from ..config import config
 from .document import _destroy_document
 from .handlers import MarkdownHandler, NotebookHandler, ScriptHandler
+from .loading import LOADING_INDICATOR_CSS_CLASS
 from .logging import LOG_SESSION_DESTROYED, LOG_SESSION_LAUNCHING
 from .state import set_curdoc, state
 
 if TYPE_CHECKING:
     from bokeh.application.application import SessionContext
     from bokeh.application.handlers.handler import Handler
+    from bokeh.document.document import Document
+
+    from ..template.base import BaseTemplate
+    from ..viewable import Viewable, Viewer
+    from .location import Location
+
+    TViewable = Viewable | Viewer | BaseTemplate
+    TViewableFuncOrPath = TViewable | Callable[[], TViewable] | os.PathLike | str
+
+
+logger = logging.getLogger('panel.io.application')
+
 
-log = logging.getLogger('panel.io.application')
+def _eval_panel(
+    panel: TViewableFuncOrPath, server_id: str, title: str,
+    location: bool | Location, admin: bool, doc: Document
+):
+    from ..pane import panel as as_panel
+    from ..template import BaseTemplate
 
+    if config.global_loading_spinner:
+        doc.js_on_event(
+            'document_ready', CustomJS(code=f"""
+            const body = document.getElementsByTagName('body')[0]
+            body.classList.remove({LOADING_INDICATOR_CSS_CLASS!r}, {config.loading_spinner!r})
+            """)
+        )
+
+    doc.on_event('document_ready', partial(state._schedule_on_load, doc))
```

Comment:
Can this cause a memory leak. In doc never being cleaned up, or am I being paranoid?

Reply:
The document gets cleaned up when the session is destroyed, including all handlers.